File: src/Product.java
import java.time.LocalDate;

public abstract class Product {
    protected String name;
    protected double price;
    protected int quantity;
    
    public Product(String name, double price, int quantity) {
        this.name = name;
        this.price = price;
        this.quantity = quantity;
    }
    
    public String getName() {
        return name;
    }
    
    public double getPrice() {
        return price;
    }
    
    public int getQuantity() {
        return quantity;
    }
    
    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }
    
    public boolean isAvailable(int requestedQuantity) {
        return quantity >= requestedQuantity;
    }
    
    public abstract boolean isExpired();
    public abstract boolean requiresShipping();
}

File: csharp/Product.cs
// csharp/Product.cs
using System;

namespace ECommerce
{
    public abstract class Product
    {
        public string Name { get; protected set; }
        public double Price { get; protected set; }
        public int Quantity { get; protected set; }

        public Product(string name, double price, int quantity)
        {
            Name = name;
            Price = price;
            Quantity = quantity;
        }

        public void SetQuantity(int quantity)
        {
            Quantity = quantity;
        }

        public bool IsAvailable(int requestedQuantity)
        {
            return Quantity >= requestedQuantity;
        }

        public abstract bool IsExpired();
        public abstract bool RequiresShipping();
    }
}
